### PR TITLE
Update SQL implementation with new features and updates Swagger

### PR DESCRIPTION
**Pull request by milismsft**
Update SQL implementation with new features and updates Swagger.
Contains everything except SqlDatabaseImpl full implementation and SqlServer create job scheduling (which will be checked in shortly).
The mocked test cases will fail because they need new recordings (will follow shortly with those as well).

### CODE REVIEW

File: src/ResourceManagement/Sql/ElasticPoolActivityImpl.cs [hunk 1]
```diff
@@ -3,137 +3,169 @@
...
 
-        ///GENMHASH:3C1909F3137E91E93C57B90768BECD1A:561DFB636B157F511DFC1C40D1BDE39E
-        public DateTime EndTime()
+        ///GENMHASH:7EC64CE674517E507F9E7D72F93A7DF6:83C9FF1F82052A9E5127DEA47990F4A0
+        public string ErrorMessage()
         {
-            return Inner.EndTime.GetValueOrDefault();
+            //$ return this.Inner.ErrorMessage();
+
+            return null;
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/ElasticPoolActivityImpl.cs]: this method contains commented java code. Is it fully implemented?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\ServiceTierAdvisorImpl.cs line 114.
```
            //$ if (this.serviceLevelObjectiveUsageMetrics == null) {
            //$ this.serviceLevelObjectiveUsageMetrics = new ArrayList<>();
            //$ foreach(var sloUsageMetricInner in this.Inner.ServiceLevelObjectiveUsageMetrics()) {
            //$ this.serviceLevelObjectiveUsageMetrics.Add(new ServiceLevelObjectiveUsageMetricImpl(sloUsageMetricInner));
            //$ }
            //$ }
            //$ return serviceLevelObjectiveUsageMetrics;

            return null;
```
Method seems to miss implementation.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\ServiceTierAdvisorImpl.cs line 181
```
        {
            //$ return this.Inner.ObservationPeriodEnd();

            return DateTime.Now;
        }
```
Please revise. Shouldn't you return inner object instead of DateTime.Now?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseBasicStorage.cs in general naming in C# are Camel case for enum values. `MAX_100_MB` should be `Max100Mb". On another thought you should name enum something like "SqlDatabaseBasicStorageSize" and then have just following:
```
        100Mb = 100L * 1024L * 1024L,
        500Mb = 500L * 1024L * 1024L,
        1Gb = 1L * 1024L * 1024L * 1024L,
        2Gb = 2L * 1024L * 1024L * 1024L
```
Max prefix everywhere is kind of redundant.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseExportRequestImpl.cs line 45 - do not throw autogenerated SDK exceptions. We will be mislead by the exception type during investigation. You should throw Exception or ArgumentNullException. Even InvalidOperationException at this point is much better than Microsoft.Rest.Azure.CloudException.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseExportRequestImpl.cs line 49
```
var storageAccountKey = (await storageAccount.GetKeysAsync())[0].Value;
```
Please first check that GetKeysAsync actually returned not null response and it has at least 1 element in the array. Otherwise caller will get nullRef exception and it is obvious what was null - array or returned value from the call.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseExportRequestImpl.cs line 61
```
                .ExportAsync(this.sqlDatabase.ResourceGroupName(), this.sqlDatabase.SqlServerName(), this.sqlDatabase.Name(), this.Inner);
```
please add cancellationToken parameter.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseForElasticPoolImpl.cs line 136
```
        ///GENMHASH:077EB7776EFFBFAA141C1696E75EF7B3:8C759FE8EBA306AE974D0A58B24683DD
        public SqlElasticPoolImpl Attach()
        {
            //$ this.sqlDatabase.AddParentDependency(this.sqlElasticPool);
            return this.sqlElasticPool;
        }
```
please revise the comment and remove if obsolete.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImportRequestImpl.cs line 54.
```
var storageAccountKey = (await storageAccount.GetKeysAsync())[0].Value;
```
same comment. Please first check that GetKeysAsync actually returned not null response and it has at least 1 element in the array. Otherwise caller will get nullRef exception and it is obvious what was null - array or returned value from the call.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImportRequestImpl.cs line 61
```
            var importExportResponseInner = await this.sqlServerManager.Inner.Databases
                .CreateImportOperationAsync(this.sqlDatabase.ResourceGroupName(), this.sqlDatabase.SqlServerName(), this.sqlDatabase.Name(), this.Inner);
```
please add cancellationToken parameter.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseMetricDefinitionImpl.cs line 49
```
        public IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISqlDatabaseMetricAvailability> MetricAvailabilities()
        {
            List<ISqlDatabaseMetricAvailability> databaseMetricAvailabilities = new List<ISqlDatabaseMetricAvailability>();
            if (this.Inner.MetricAvailabilities != null)
            {
                foreach (var metricAvailability in this.Inner.MetricAvailabilities)
                {
                    databaseMetricAvailabilities.Add(new SqlDatabaseMetricAvailabilityImpl(metricAvailability));
                }
            }

            return databaseMetricAvailabilities.AsReadOnly();
        }
```

please cache the readonly list and next time just return it instead of recreating.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseMetricImpl.cs, Line 54, - please cache the readonly list and next time just return it instead of recreating.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabasePremiumStorage.cs, same comment about enum values - Max prefix everywhere is kind of redundant. Please revise.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabasesAsExternalChildResourcesImpl.cs is not implemented.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseStandardStorage.cs, same comment about enum values - Max prefix everywhere is kind of redundant. Please revise.

**Comment by hovsepm** [comment]
ResourceManagement\Sql\SqlElasticPoolBasicEDTUs.cs, same comment about enum values - eDTU prefix everywhere is kind of redundant. Please revise.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolBasicMaxEDTUs.cs,  eDTU prefix everywhere is kind of redundant. Please revise. Any point of having a single value enum? Can't we default it in code?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolBasicMinEDTUs.cs, eDTU prefix everywhere is kind of redundant. Please revise.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolForDatabaseImpl.cs, line 161
```
        public SqlDatabaseImpl Attach()
        {
            //$ this.sqlDatabase.AddParentDependency(this.sqlElasticPool);
            return this.sqlDatabase;
        }
```
please verify comment in the code and if obsolete - remove.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolImpl.cs, line 272,
```
        /////GENMHASH:7A9E1ACC5D9B5ED3EF93A2ABFD978F14:1F4B6D4F1D6BA088BF475B965981DC9B
        //internal void AddParentDependency(IHasTaskGroup parentDependency)
        //{
        //    //$ this.AddDependency(parentDependency);
        //    //$ }

        //}
```
please verify comment in the code and if obsolete - remove.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolImpl.cs, line 287,
```
        ///GENMHASH:9557699E7EE892CCCC89A074E0915333:27E486AB74A10242FF421C0798DDC450
        public void BeforeGroupCreateOrUpdate()
        {
            //$ 

        }
```

what is the point of having an empty method?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolImpl.cs, line 417
```
        public IUpdate Update()
        {
            //$ super.PrepareUpdate();
            return this;
        }
```
please verify comment in the code and if obsolete - remove.

File: src/ResourceManagement/Sql/ElasticPoolActivityImpl.cs (hunk 1 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/ElasticPoolActivityImpl.cs]: fixed

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolsAsExternalChildResourcesImpl.cs is not implemented.

**Comment by hovsepm** [comment]
\src\ResourceManagement\Sql\SqlElasticPoolStandardStorage.cs, - Max prefix everywhere is kind of redundant. Please revise.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlFirewallRulesAsExternalChildResourcesImpl.cs is not implemented

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlServerImpl.cs is partially implemented. Please revise.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlWarehouseImpl.cs line 35
```
        public async Task PauseDataWarehouseAsync(CancellationToken cancellationToken = default(CancellationToken))
        {
            await this.sqlServerManager.Inner.Databases
                .PauseAsync(this.ResourceGroupName(), this.SqlServerName(), this.Name());
        }
```

pass in cancelationToken.

**Comment by hovsepm** [comment]
@milismsft please pull from upstream and resolve merge conflicts.

**Comment by hovsepm** [comment]
Samples\samples.json, line 240."restorr"... rrrr? too angry? :)))

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\RecommendedElasticPoolImpl.cs, ine 29
```
        public IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISqlDatabase> Databases()
        {
            List<ISqlDatabase> databases = new List<ISqlDatabase>();
            if (this.Inner.Databases != null)
            {
                foreach (var databaseInner in this.Inner.Databases)
                {
                    databases.Add(new SqlDatabaseImpl(databaseInner.Name, this.sqlServer, databaseInner, this.Manager()));
                }
            }

            return databases.AsReadOnly();
        }
```

The list should be cached.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\ServiceTierAdvisorImpl.cs, line 94:
```
        public IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISloUsageMetricInterface> ServiceLevelObjectiveUsageMetrics()
        {
            if (this.sloUsageMetrics == null)
            {
                this.sloUsageMetrics = new List<ISloUsageMetricInterface>();
                if (this.Inner.ServiceLevelObjectiveUsageMetrics != null)
                {
                    foreach (var sloUsageMetricInner in this.Inner.ServiceLevelObjectiveUsageMetrics)
                    {
                        this.sloUsageMetrics.Add(new SloUsageMetricImpl(sloUsageMetricInner));
                    }
                }
            }

            return this.sloUsageMetrics.AsReadOnly();
        }
```
The list should be cached.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\ServiceTierAdvisorImpl.cs, line 112
```
        public IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.IServiceLevelObjectiveUsageMetric> ServiceLevelObjectiveUsageMetric()
        {
            if (this.serviceLevelObjectiveUsageMetrics == null)
            {
                this.serviceLevelObjectiveUsageMetrics = new List<IServiceLevelObjectiveUsageMetric>();
                foreach (var sloUsageMetricInner in this.Inner.ServiceLevelObjectiveUsageMetrics)
                {
                    this.serviceLevelObjectiveUsageMetrics.Add(new ServiceLevelObjectiveUsageMetricImpl(sloUsageMetricInner));
                }
            }
            return this.serviceLevelObjectiveUsageMetrics.AsReadOnly();
        }
```
The list should be cached.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseExportRequestImpl.cs, line 49:
```
var storageKeys = (await storageAccount.GetKeysAsync());
```
pass cancellationToken.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseExportRequestImpl.cs, line 62:
```
await cloudBlobClient.GetContainerReference(containerName).CreateIfNotExistsAsync();
```
cancellationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImpl.cs, line 294:
```
        public SqlDatabaseImpl WithExistingSqlServer(ISqlServer sqlServer)
        {
            if (sqlServer == null)
            {
                throw new ArgumentNullException("sqlServer");
            }
            return this;
        }
```

Why do you even need this method with the parameter if nothing is done inside the method? do you want to keep reference of sqlServer for further use?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImpl.cs, 902:
```
                var dbInner = this.sqlServerManager.Inner.Databases
                    .ImportAsync(this.resourceGroupName, this.sqlServerName, this.importRequestInner, cancellationToken);
```
missing `await`.

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImpl.cs, line 956,
```
return await this.CreateResourceAsync();
```
cancellationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImpl.cs, line 975
```
await this.DeleteResourceAsync();
```
cancellationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseImportRequestImpl.cs, line 54:
```
var storageKeys = (await storageAccount.GetKeysAsync());
```
cancelationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseMetricImpl.cs, line 54
```
        public IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISqlDatabaseMetricValue> MetricValues()
        {
            List<ISqlDatabaseMetricValue> sqlDatabaseMetricValues = new List<ISqlDatabaseMetricValue>();
            if (this.Inner.MetricValues != null)
            {
                foreach (var metricValue in this.Inner.MetricValues)
                {
                    sqlDatabaseMetricValues.Add(new SqlDatabaseMetricValueImpl(metricValue));
                }
            }
            //$ foreach(var metricValue in this.Inner.MetricValues()) {
            //$ sqlDatabaseMetricValues.Add(new SqlDatabaseMetricValueImpl(metricValue));
            //$ }
            //$ }
            //$ return Collections.UnmodifiableList(sqlDatabaseMetricValues);

            return sqlDatabaseMetricValues.AsReadOnly();
        }
```
This does not seems to be cached and also please remove commented java code or comment your commented code :)

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseThreatDetectionPolicyImpl.cs, line 83:
```
        public IUpdate Update()
        {
            return this;
        }
```
update is no-op?

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlDatabaseThreatDetectionPolicyImpl.cs, 101
```
return await CreateResourceAsync();
```
cancellationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolImpl.cs, line 412:
```
        public IUpdate Update()
        {
            // Future reference when enabling the proper framework which will track dependencies 
            //$ super.PrepareUpdate();
            return this;
        }
```
no-op update is kind of suspicious...

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlElasticPoolStandardStorage.cs please fix enum values to MaxBlahBlah

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlRestorableDroppedDatabaseImpl.cs, line 36:
```
.GetAsync(this.resourceGroupName, this.sqlServerName, this.Inner.Id);
```
cancellationToken

**Comment by hovsepm** [comment]
src\ResourceManagement\Sql\SqlServerImpl.cs, line 411
```
//            List<Task> deleteWorkers = new List<Task>();
```
remove dead comment.

**Pull request closed without merging**